File: backend/src/test/java/com/checklistboteco/backend/WorkClockCalculatorTest.java
package com.checklistboteco.backend;

import com.checklistboteco.backend.model.Models.WorkClockEntry;
import com.checklistboteco.backend.workclock.domain.WorkClockCalculator;
import com.checklistboteco.backend.workclock.domain.WorkClockModels.UserWorkSchedule;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.List;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class WorkClockCalculatorTest {
    @Test
    void eightWorkedHoursAndOneBreakHourCompletesRegularDay() {
        List<WorkClockEntry> entries = List.of(
            entry("ENTRADA", 8),
            entry("ALMOCO_INICIO", 12),
            entry("ALMOCO_FIM", 13),
            entry("SAIDA", 17)
        );
        var summary = WorkClockCalculator.summarizeDay(entries, 0L);
        assertEquals(8L * 60 * 60 * 1000, summary.workedMillis());
        assertEquals(1L * 60 * 60 * 1000, summary.lunchMillis());
        assertEquals(0L, summary.missingDailyMillis());
        assertEquals(0L, summary.missingBreakMillis());
    }

    @Test
    void twelveHourShiftRequiresTwoBreakHours() {
        List<WorkClockEntry> entries = List.of(
            entry("ENTRADA", 6),
            entry("ALMOCO_INICIO", 12),
            entry("ALMOCO_FIM", 13),
            entry("SAIDA", 19)
        );
        var summary = WorkClockCalculator.summarizeDay(entries, 0L);
        assertEquals(12L * 60 * 60 * 1000, summary.workedMillis());
        assertTrue(summary.requiresTwoHoursRest());
        assertEquals(1L * 60 * 60 * 1000, summary.missingBreakMillis());
    }

    @Test
    void overtimeIsOnlyAboveFortyWeeklyHours() {
        long fortyHours = 40L * 60 * 60 * 1000;
        assertEquals(0L, WorkClockCalculator.overtimeMillis(fortyHours));
        assertEquals(1L * 60 * 60 * 1000, WorkClockCalculator.overtimeMillis(fortyHours + 3_600_000L));
    }

    @Test
    void absenceCountsScheduledDayWithoutEntrada() {
        UserWorkSchedule schedule = new UserWorkSchedule();
        schedule.workingDaysOfWeek = List.of(1, 2, 3, 4, 5);
        LocalDate monday = LocalDate.of(2026, 6, 15);
        int absences = WorkClockCalculator.countAbsenceDays(List.of(), schedule, monday, monday);
        assertEquals(1, absences);
    }

    @Test
    void absenceDetailsListMissingEntryAndMissingExitForPastDays() {
        UserWorkSchedule schedule = new UserWorkSchedule();
        schedule.workingDaysOfWeek = List.of(1, 2, 3, 4, 5, 6);
        LocalDate friday = LocalDate.of(2026, 6, 19);
        LocalDate saturday = LocalDate.of(2026, 6, 20);
        WorkClockEntry saturdayEntry = entry("ENTRADA", saturday, 8);

        var absences = WorkClockCalculator.absenceDetails(List.of(saturdayEntry), schedule, friday, saturday);

        assertEquals(2, absences.size());
        assertEquals("2026-06-19", absences.get(0).date);
        assertEquals("Sem entrada registrada", absences.get(0).reason);
        assertEquals("2026-06-20", absences.get(1).date);
        assertEquals("Entrada sem saída registrada", absences.get(1).reason);
    }

    private WorkClockEntry entry(String type, int hour) {
        return entry(type, LocalDate.of(2026, 6, 20), hour);
    }

    private WorkClockEntry entry(String type, LocalDate date, int hour) {
        WorkClockEntry entry = new WorkClockEntry();
        entry.type = type;
        entry.registeredAt = date.atTime(hour, 0).atZone(ZoneId.of("America/Sao_Paulo")).toInstant().toEpochMilli();
        return entry;
    }
}

File: composeApp/src/commonTest/kotlin/com/checklistboteco/domain/model/WorkClockCalculatorTest.kt
package com.checklistboteco.domain.model

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class WorkClockCalculatorTest {
    @Test
    fun eightWorkedHoursAndOneBreakHourCompletesRegularDay() {
        val entries = listOf(
            entry(WorkClockType.ENTRADA, hour = 8),
            entry(WorkClockType.ALMOCO_INICIO, hour = 12),
            entry(WorkClockType.ALMOCO_FIM, hour = 13),
            entry(WorkClockType.SAIDA, hour = 17)
        )

        val summary = WorkClockCalculator.summarizeDay(entries)

        assertEquals(8.hours, summary.workedMillis)
        assertEquals(1.hours, summary.lunchMillis)
        assertEquals(0L, summary.missingDailyMillis)
        assertEquals(0L, summary.missingBreakMillis)
        assertEquals(0L, summary.breakOverageMillis)
    }

    @Test
    fun missingDailyHoursAreCalculatedFromWorkedTimeNotStartTime() {
        val entries = listOf(
            entry(WorkClockType.ENTRADA, hour = 11),
            entry(WorkClockType.SAIDA, hour = 17)
        )

        val summary = WorkClockCalculator.summarizeDay(entries)

        assertEquals(6.hours, summary.workedMillis)
        assertEquals(2.hours, summary.missingDailyMillis)
        assertEquals(1.hours, summary.missingBreakMillis)
    }

    @Test
    fun twelveHourShiftRequiresTwoBreakHours() {
        val entries = listOf(
            entry(WorkClockType.ENTRADA, hour = 6),
            entry(WorkClockType.ALMOCO_INICIO, hour = 12),
            entry(WorkClockType.ALMOCO_FIM, hour = 13),
            entry(WorkClockType.SAIDA, hour = 19)
        )

        val summary = WorkClockCalculator.summarizeDay(entries)

        assertEquals(12.hours, summary.workedMillis)
        assertTrue(summary.requiresTwoHoursRest)
        assertEquals(1.hours, summary.missingBreakMillis)
    }

    @Test
    fun entryIsNotLateWithoutFixedSchedule() {
        assertFalse(
            WorkClockCalculator.isLateEntry()
        )
    }

    private fun entry(type: WorkClockType, hour: Int): WorkClockEntry {
        return WorkClockEntry(
            id = hour.toLong(),
            userId = 1L,
            type = type,
            registeredAt = hour.hours,
            location = WorksiteLocation.point,
            distanceFromWorkMeters = 0.0,
            isLate = false
        )
    }

    private val Int.hours: Long
        get() = this * 60L * 60L * 1000L
}
